### PR TITLE
Added new IBAN uitype

### PR DESCRIPTION
**Pull request by konadrian**
To test this add a field with 329 uitype and add to the fieldparams 
$fieldParams = ['country' => 'PL', 'sortCode' => '10500161', 'clientId' => '5555', 'conditions' => [
	'legal_form:PLL_NATURAL_PERSON:pesel',
	'defaultValue:vat_id'
], 'clientIdPattern' => 'NNNNNNNNNNNN'];

{"country":"PL","sortCode":"10500161","clientId":"5555","conditions":["legal_form:PLL_NATURAL_PERSON:pesel","defaultValue::vat_id"],"clientIdPattern":"NNNNNNNNNNNN"}

### CODE REVIEW

File: modules/Vtiger/uitypes/IBAN.php [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+		$fieldName = $this->getFieldModel()->getName();
+		$this->recordModel = $recordModel;
+		if (!$requestFieldName) {
+			$requestFieldName = $fieldName;
+		}
+		$value = $request->getByType($requestFieldName, 'Text');
+		if (!$value) {
+			$value = $this->getIBANValue();
+		}
+		var_dump($value);
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/IBAN.php]: ?

File: modules/Vtiger/uitypes/IBAN.php [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+	 * Checks if field params has mandatory values.
+	 *
+	 * @return bool
+	 */
+	protected function checkIfFieldHasMandatoryParams(): bool
+	{
+		if ($this->fieldParams = $this->getFieldModel()->getFieldParams()) {
+			return isset($this->fieldParams['country']) && isset($this->fieldParams['sortCode'], $this->fieldParams['clientId'], $this->fieldParams['conditions']);
+		}
+		return false;
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/IBAN.php]: ?

File: modules/Vtiger/uitypes/IBAN.php [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+	{
+		$clientPattern = $this->fieldParams['clientIdPattern'];
+		$patternLength = \strlen($clientPattern);
+
+		$payerId = '';
+		$letters = range('A', 'Z');
+		$numbers = range('0', '9');
+		$lettersAndNumbers = array_merge($letters, $numbers);
+		for ($charPosition = 0; $charPosition < $patternLength; ++$charPosition) {
+			$clientPattern[$charPosition];
```
Comment (rskrzypczak) [inline modules/Vtiger/uitypes/IBAN.php]: ?

File: app/Conditions/QueryFields/IBANField.php [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+ * IBAN query field file.
+ *
+ * @package   App
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Adrian Kon <a.kon@yetiforce.com>
+ */
+
+namespace App\Conditions\QueryFields;
```
Comment (rskrzypczak) [inline app/Conditions/QueryFields/IBANField.php]: Add for App\Conditions\RecordFields to

**Review by rskrzypczak — CHANGES_REQUESTED** [review]

File: composer.json [hunk 5]
```diff
@@ -83,7 +83,8 @@
     "phpmailer/dkimvalidator": "^v0.3",
     "league/climate": "^3.6",
     "twig/twig": "^3.0",
-    "illuminate/support": "^8.0"
+    "illuminate/support": "^8.0",
+    "globalcitizen/php-iban": "4.0.0"
```
Comment (mariuszkrzaczkowski) [inline composer.json:87]:
```suggestion
    "globalcitizen/php-iban": "^4.0.0"
```

File: app/Fields/Iban.php [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license   YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author    Adrian Kon <a.kon@yetiforce.com>
+ */
+
+namespace App\Fields;
+
+/**
+ * Iban class.
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Iban.php:15]: ?

File: app/Fields/Iban.php [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+{
+	/**
+	 * Function returns IBAN value.
+	 *
+	 * @param array $fieldParams
+	 * @param array $recordData
+	 *
+	 * @return string
+	 */
+	public function getIBANValue(array $fieldParams, array $recordData): string
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Iban.php:27]: getIBANValue ?

$recordData  >>  $recordModel

File: app/Fields/Iban.php [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+	{
+		$this->fieldParams = $fieldParams;
+		$this->recordData = $recordData;
+		if ($this->checkIfFieldHasMandatoryParams()) {
+			$iban = new \PHP_IBAN\IBAN();
+			$this->countryIBAN = new \PHP_IBAN\IBANCountry($this->fieldParams['country']);
+			$countryIBANLength = $this->countryIBAN->IBANLength();
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Iban.php:34]: countryIBANLength  >> countryIbanLength

File: app/Fields/Iban.php [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+			$checkSumValueForCreateIBAN = '00';
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Iban.php:35]: checkSumValueForCreateIBAN  > checkSumValueForCreateIban

File: app/Fields/Iban.php [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+	}
+
+	/**
+	 * Checks if field params has mandatory values.
+	 *
+	 * @return bool
+	 */
+	protected function checkIfFieldHasMandatoryParams(): bool
+	{
+		return isset($this->fieldParams['country']) && isset($this->fieldParams['sortCode'], $this->fieldParams['clientId'], $this->fieldParams['conditions']);
```
Comment (mariuszkrzaczkowski) [inline app/Fields/Iban.php:54]: ?

File: install/install_schema/scheme.sql [hunk 11]
```diff
@@ -1,3 +1,8 @@
+/*
+SQLyog Ultimate v13.0.0 (64 bit)
```
Comment (mariuszkrzaczkowski) [inline install/install_schema/scheme.sql:2]: ???

File: modules/Vtiger/handlers/AutoFillIban.php [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+ * Auto fill iban class.
+ *
+ * @package Handler
+ *
+ * @copyright YetiForce Sp. z o.o
+ * @license YetiForce Public License 3.0 (licenses/LicenseEN.txt or yetiforce.com)
+ * @author Adrian Kon <a.kon@yetiforce.com>
+ */
+/**
+ * Vtiger_AutoFillIban_Handler class.
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:13]: ?

File: modules/Vtiger/handlers/AutoFillIban.php [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+ */
+class Vtiger_AutoFillIban_Handler
+{
+	/**
+	 * EntityBeforeSave function.
+	 *
+	 * @param App\EventHandler $eventHandler
+	 */
+	public function entityBeforeSave(App\EventHandler $eventHandler)
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:22]: public function entityBeforeSave(App\EventHandler $eventHandler): void

File: modules/Vtiger/handlers/AutoFillIban.php [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+	{
+		$recordModel = $eventHandler->getRecordModel();
+		$ibanUiType = 329;
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:25]: static

File: modules/Vtiger/handlers/AutoFillIban.php [hunk 15]
```diff
@@ -0,0 +1,39 @@
...
+		$allIbanFields = $recordModel->getModule()->getFieldsByUiType($ibanUiType);
+		if ($allIbanFields) {
+			$recordData = $recordModel->getData();
+			foreach ($allIbanFields as $moduleIbanField) {
+				if (!$moduleIbanField->hasDefaultValue() && !$recordModel->get($moduleIbanField->getName())) {
+					$ibanField = new \App\Fields\Iban();
+					$fieldParams = $moduleIbanField->getFieldParams();
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:32]: ?

File: modules/Vtiger/handlers/AutoFillIban.php [hunk 16]
```diff
@@ -0,0 +1,39 @@
...
+					$ibanValue = $ibanField->getIBANValue($fieldParams, $recordData);
```
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:33]: ?

File: app/Fields/Iban.php (hunk 10 above)
Comment (mariuszkrzaczkowski) [reply app/Fields/Iban.php:54]: throw Exceptions

File: modules/Vtiger/handlers/AutoFillIban.php (hunk 16 above)
Comment (mariuszkrzaczkowski) [inline modules/Vtiger/handlers/AutoFillIban.php:33]: ?

**Pull request merged**